### ISSUE TITLE
Transfer Unicode ICU modules to Node.js

### ISSUE THREAD

**Issue by obensource** [nodejs/admin#532]
## Request Description
@srl295 and other Unicode Consortium members who maintain the ICU npm modules that node consumes have expressed an intent to transfer ownership of the following repositories to the nodejs org:
* [`full-icu-npm`](https://github.com/unicode-org/full-icu-npm)
* [`full-icu-test`](https://github.com/unicode-org/full-icu-test)
* [`icu4c-data-npm`](https://github.com/unicode-org/icu4c-data-npm)

These modules deliver [full-ICU support to node](https://nodejs.org/api/intl.html#intl_embed_the_entire_icu_full_icu). If transferred into this org, their repositories will become primarily maintained by the @nodejs/i18n-api team.

## Why is this necessary?
Node is their main target, and though @srl295 has been championing these modules for some time – his priorities recently changed and he will be more solely focused on developing [CLDR](https://github.com/unicode-org/cldr). He has expressed the intent to remain available for questions regarding ICU issues from @nodejs/i18n-api, @nodejs/i18n-api, and anyone else.

## What does this mean?
With Steven focused on CLDR, the @nodejs/i18n team will be seeking to onboard new globalization contributors who are interested in maintaining ICU in node. We're planning to schedule open mentorship sessions with @srl295 (and any other node or Unicode Org contributors who wish to join) in order to make the onboarding process as effective and enjoyable as possible. 

**As an outcome**: we hope to foster more reciprocity for g11n in node.

**Comment by srl295** [comment]
- `full-icu-npm` is the `full-icu` npm module
- `full-icu-test` is the CITGM test for it. 
- `icu4c-data-npm` is the helper which builds and publishes icu4c data to NPM. This is becoming obsolete whenever https://github.com/unicode-org/full-icu-npm/issues/36 is implemented, because on the ICU side the data is published as seaprate zip files in github https://unicode-org.atlassian.net/browse/ICU-20600


So, this could be a good candidate for a monorepo.

**Comment by bnb** [comment]
I've got a minor concern here that since @srl295's focus has changed away from these modules and there was no mention of anyone _already_ maintaining it (looking to onboard vs. enabling existing maintainers) that there's a non-trivial chance that we won't be able to maintain these effectively.

**Comment by ryzokuken** [comment]
:raised_hand_with_fingers_splayed: I could help out this these packages if you need any.

**Comment by obensource** [comment]
@bnb out of the gate, if not much action is taken other than to transfer these to the node org – there's definitely a chance we won't be able to maintain them effectively. 💯

A couple thoughts though:
* Before transferring ownership, we should discuss this in full at our next @nodejs/i18n WG meeting with @srl295 and any other active maintainers of these repos in order to better understand the requirements of effectively maintaining them – and if members of node or Unicode are better suited to do this. There's really no other org these are directly related to, so coming to a conclusion on this will be key.
* As mentioned, this request is coming at a time where node's central g11n contributor is shifting focus – and a gap will be felt. So we're going to need to put out a call for participation soon for the @nodejs/i18n-api team regardless, which will require better documentation for new contributor onboarding, etc. @srl295 will be available to guide us as we shift focus on building sustainability for this, so that's very positive. I'm excited to support this effort, and imo having the modules live in this org could be a good forcing function to expand and diversify participation for g11n in node (everything under the same roof). However, if after further discussion the maintainers express that onboarding new contributors through node to help maintain these modules might be less effective than keeping the repos in Unicode and doing something similar there: then we can defer to their judgement and support them as they need it. 👍

Thanks a mil @bnb. 🙏

@srl295 any thoughts?

**Comment by srl295** [comment]
> I've got a minor concern here that since @srl295's focus has changed away from these modules and there was no mention of anyone _already_ maintaining it (looking to onboard vs. enabling existing maintainers) that there's a non-trivial chance that we won't be able to maintain these effectively.

Really great thoughts. However, it's not getting any maintenance by other ICU people. (Historically it started as part of ICU because I was revving the ICU tools, but that's been superseded by other work within Icu.) And the reality is that it is commonly used within the node ecosystem. So, owner wise if it's within node i18n it will be more likely to get any updates. 

Icu-tc was supportive of this when discussed last.

**Comment by srl295** [comment]
Also I'm glad to help onboard anyone/everyone, @ryzokuken (Thanks) or others.

**Comment by bnb** [comment]
+1 then. I would like to only request that if we do accept that we also have a plan for messaging in the case where we're not able to find active maintainers / the maintainers that start are not able to commit to long-term maintenance.

**Comment by jasnell** [comment]
+1 from me and many thanks to @srl295 for keeping it going thus far

**Comment by mhdawson** [comment]
+1 from me and thanks for all of the hard work put in by @srl295

**Comment by srl295** [comment]
I'm going to start this up again.

**Comment by srl295** [comment]
would include these two packages also:
- https://www.npmjs.com/package/full-icu
- https://www.npmjs.com/package/icu4c-data

**Comment by srl295** [comment]
@jasnell @mhdawson @obensource  et al, I have formally requested this transfer on the Unicode side.

**Comment by srl295** [comment]
@bnb / @mhdawson : unicode has approved this! What's next?

**Comment by mcollina** [comment]
@ryzokuken are you still up for taking ownership of those modules?

**Comment by obensource** [comment]
@srl295 @mhdawson @jasnell wonderful news! Thank you for your amazing work Steven! 👏👏👏

I'm available to help with the transfer – @mhdawson @mcollina please let me know here or feel free to reach out with what steps need to be taken.

@mcollina @ryzokuken I'd love to help maintain these modules as well. 🚀

**Comment by ryzokuken** [comment]
@mcollina yes!

@obensource absolutely, would love to have your help :D

**Comment by obensource** [comment]
@ryzokuken great! I'll reach out and let's sync soon! 🍻

**Comment by srl295** [comment]
@ryzokuken @obensource SGTM. I can do the actual initiation on the Unicode side.

**Comment by obensource** [comment]
@srl295 wonderful – thank you for doing that and if I don't hear from the node folks we've pinged here soon I'll reach out directly. 👍

**Comment by srl295** [comment]
I did a minor update of full-icu to rollup some pending PRs, and I also made sure the metadata, copyright, license on everything is up to date.

**Comment by obensource** [comment]
Thanks a million for bringing it up to date @srl295! 😎🙌

**Comment by mhdawson** [comment]
@obensource from what I remember we need to make sure the repos have the following files:

* CONTRIBUTING.md  with DCO in it
* CODE_OF_CONDUCT.md
* LICENSE  (licence needs to be compatible with OpenJS IP policy approved licences or we need to request an exception)Once

Once we've validated that then I think we need to make one of the people in the project who have admin rights an admin on the repos and they can do the transfer.

**Comment by mhdawson** [comment]
One more thing, the github templates should also be set up so that the DCO is displayed when a PR is created as we do that in the core node repo as well.

**Comment by ryzokuken** [comment]
@obensource let me know if you require any assistance :D

**Comment by srl295** [comment]
> @obensource from what I remember we need to make sure the repos have the following files:
> 
> 
> 
> * CONTRIBUTING.md  with DCO in it
> 
> * CODE_OF_CONDUCT.md
> 
> * LICENSE  (licence needs to be compatible with OpenJS IP policy approved licences or we need to request an exception)Once
> 
> 
> 
> Once we've validated that then I think we need to make one of the people in the project who have admin rights an admin on the repos and they can do the transfer.

Want to just make PRs to request those changes?

I can initiate the transfer whenever needed

**Comment by srl295** [comment]
What's the latest on this?

**Comment by bnb** [comment]
CC @obensource got an update?

**Comment by obensource** [comment]
@srl295 @bnb thanks for your patience, I've blocked out some time to work on this today / this week. @srl295 I'll ping you as I go and let's shoot to get the transfer finalized this week! 🎉

**Comment by obensource** [comment]
@srl295 @bnb adding the missing docs now + adding the github templates that @mhdawson mentioned "should also be set up so that the DCO is displayed when a PR is created".

I'll have some PRs in for your review in a bit @srl295. 👍

**Comment by srl295** [comment]
@obensource you may already be doing this, but please link them here (can probably just reference this issue) for reference

**Comment by srl295** [comment]
> would include these two packages also:
> 
> * https://www.npmjs.com/package/full-icu
> * https://www.npmjs.com/package/icu4c-data

what's the procedure for changing NPM module ownership?

**Comment by obensource** [comment]
@srl295 you got it, forked and working on them here at the moment:
* [full-icu-npm](https://github.com/obensource/full-icu-npm)
* [full-icu-test](https://github.com/obensource/full-icu-test)
* [icu4c-data-npm](https://github.com/obensource/icu4c-data-npm)

I'm currently adding the docs, and I will submit PRs to the upstream repos for each.

**Comment by obensource** [comment]
As far as changing module ownership, I'm running with what @mhdawson said at the moment as far as the repositories are concerned: 

> Once we've validated that then I think we need to make one of the people in the project who have admin rights an admin on the repos and they can do the transfer.

Regarding the packages, I haven't looked into it yet – but I'm sure there's a way to do [this type of transfer] (https://docs.npmjs.com/transferring-a-package-from-a-user-account-to-another-user-account) between the orgs as long as it's facilitated by people who have the admin privileges. I'm sure @mhdawson @bnb could tell us in the meantime if we don't get to it first. 😄

**Comment by srl295** [comment]
Ah right, I am glad to simply *add* https://www.npmjs.com/~obensource as a maintainer if that is sufficient?

**Comment by obensource** [comment]
@srl295 that should be fine! Thanks a million Steven! 🙌

**Comment by srl295** [comment]
^ added you as maintainer

**Comment by obensource** [comment]
@srl295 So here's where rubber meets the road: I'm blowing away the [Unicode License](https://github.com/obensource/full-icu-npm/blob/master/LICENSE), and adding the [Node.js License](https://github.com/nodejs/node/blob/master/LICENSE) – which transitions it to MIT, but also keeps the ICU License, and provisions for locale data (of which you added to the Node.js license). Does that sound correct, or in your opinion – is there a strict set of licenses that this should be subject to as owned by the node-org?

**Comment by srl295** [comment]
> @srl295 So here's where rubber meets the road: I'm blowing away the [Unicode License](https://github.com/obensource/full-icu-npm/blob/master/LICENSE), and adding the [Node.js License](https://github.com/nodejs/node/blob/master/LICENSE) – which transitions it to MIT, but also keeps the ICU License, and provisions for locale data (of which you added to the Node.js license). Does that sound correct, or in your opinion – is there a strict set of licenses that this should be subject to as owned by the node-org?

That might be OK, but maybe it would be better/safer to rename LICENSE to `unicode-license.txt` and just note somewhere, "portions from `full-icu` from Unicode, see unicode-license.txt" ? Then it kind of covers both the old and new world, and then deduplicate the license once it lands on the Node side?

**Comment by obensource** [comment]
@srl295 That sounds good to me for now – I think for our sake it covers each side of the fence. If there's an issue with that, please feel free to advise us @mhdawson @bnb, et al. 👍

**Comment by obensource** [comment]
@srl295 Adding MIT makes this repository comply with the [OpenJSF IP Policy](https://openjsf.org/wp-content/uploads/sites/84/2019/10/OpenJS-Foundation-IP-Policy-2019-10-22.pdf).

The only thing I'm uncertain about is if retaining any existing Unicode licenses as well makes it subject to what @mhdawson called an 'exception':

> "license needs to be compatible with OpenJS IP policy approved licenses or we need to request an exception"

**Comment by obensource** [comment]
I don't think so, but please advise @mhdawson. 😄

**Comment by obensource** [comment]
@srl295 @mhdawson @bnb et al. The PRs are in, please review at your convenience. 🙏

* https://github.com/unicode-org/full-icu-npm/pull/58
* https://github.com/unicode-org/full-icu-test/pull/5
* https://github.com/unicode-org/icu4c-data-npm/pull/10

Thanks a million all! 🎉

**Comment by srl295** [comment]
Approved… would be nice to get a review from  @mhdawson ||  @bnb before merging and initiating transfer?

**Comment by srl295** [comment]
> @srl295 Adding MIT makes this repository comply with the [OpenJSF IP Policy](https://openjsf.org/wp-content/uploads/sites/84/2019/10/OpenJS-Foundation-IP-Policy-2019-10-22.pdf).
> 
> The only thing I'm uncertain about is if retaining any existing Unicode licenses as well makes it subject to what @mhdawson called an 'exception':
> 
> > "license needs to be compatible with OpenJS IP policy approved licenses or we need to request an exception"

well, icu itself has been pulled into node before many times… and,  `Unicode-DFS-2016 ` is OSI approved:

https://github.com/unicode-org/full-icu-npm/blob/master/package.json#L16

**Comment by obensource** [comment]
@srl295 100%. We should golden then! ☀️😎

**Comment by obensource** [comment]
@srl295 do I need to be added as a Unicode Consortium member (or be given write access to these repositories) in order to merge the changes, or will someone else from the UC need to review and then merge? No problem either way, and thanks! 😄

**Comment by srl295** [comment]
> @srl295 do I need to be added as a Unicode Consortium member (or be given write access to these repositories) in order to merge the changes, or will someone else from the UC need to review and then merge? No problem either way, and thanks! 😄

I'll merge

**Comment by srl295** [comment]
All transfers requested! 
I set the team owners to:  collaborators + i18n .. please fix as needed

**Comment by obensource** [comment]
@srl295 will do, thanks a million! 🙌

@mhdawson @bnb, et al: if there are no issues with the current state of this transfer, I consider this issue to be closed. 👍

Thanks everyone! 🍻

**Comment by obensource** [comment]
## Closing the issue
Given that all of the three repositories to be transferred have now been transferred and meet the governance requirements of Node.js & the OpenJSF, I'm closing this issue.

**Welcome to node, ICU modules! 👏👏👏**

<img src="https://user-images.githubusercontent.com/1910457/117685981-460ff500-b16b-11eb-86a5-e18fa0b3b7ee.gif" width="350" />

## New Locations
New locations of the repos:
* [nodejs/full-icu-npm](https://github.com/nodejs/full-icu-npm)
* [nodejs/full-icu-test](https://github.com/nodejs/full-icu-test)
* [nodejs/icu4c-data-npm](https://github.com/nodejs/icu4c-data-npm)

## Thank you! 🙏
Thanks a million to everyone who made this transition possible (@srl295, @mhdawson, @bnb, @ryzokuken, @jasnell, @mcollina). Y'all are the best! 🍻
